### DISCUSSION TITLE
Before you go any further, learn Ada.

### DISCUSSION THREAD

**Discussion by jmccabe** [carbon-language/carbon-lang#1720] [Language design]
The US DoD spent loads of money on the design and implementation of the Ada language, which was released for public use just under 40 years ago. Its design solved numerous problems inherent in the existing languages of the time, many of which continue in C++ and the languages that try to retain C++ style syntax.

I'm not suggesting that you abandon the project and use Ada instead (although that probably wouldn't be a bad idea) but, having used Ada myself for a number of years before moving to C++, I find it shocking that the C++ community seems to be so happy to get half-baked new features added to the language rather than sorting out the fundamental inadequacies of the core language. 

There are a number of Ada features that just make your job as a software developer easier, more reliable and clearer. The ones I particularly care about are:

**Enumerated types.**

In particular:

1. Differentiation between value/representation, and position
2. Being able to use them as the index type in an array (see also "Arrays") 
3. Being able to reference them using attributes, e.g. 'first. ' last, etc.
4. 'image providing a string representation of the enumeration literal

C++''s enum class is such a dreadful let-down! 

Any C/C++ programmer ought to recognise at least some of these as being things that require messing about with legacy constructs to achieve. Programmers shouldn't need to put up with that nonsense in this day and age.

**Arrays**

Definition of user-defined array types that are incompatible with each other, i.e. strong typing, not type aliases.

Definition of array types where the start index is NOT zero; "off-by-one" remains a huge problem when dealing with arrays, for no other reason than C arrays (and, hence C++ and multiple other languages) are like this because it made the compilers easier to write.

Arrays indexed by any discrete type (i.e. Including enumerated types).

Being able to use properties of an array, e.g. value of start index, value of last index, range etc. 

std::array in C++ is such a dreadful let-down!

**Named Parameter Association**

Obvious; allow parameter names to be given at a function call site to aid readability, and to improve the flexibility of defaulted parameter values (no need for them to be at the end of the list, and you can genuinely skip ones you just want to use the default of)

**User-defined Numeric Types that ARE NOT multiples of 1 byte**

The language should support types that genuinely represent the range of a property without the programmer needing to mess about trying to check etc

This includes wrap-semantics on arbitrary width types ("modular types" in Ada.

**Reliable And Consistent Representation Definitions**

This is really more geared at embedded systems but, since C++ is quite widely used in those, it should be easy to define portable representations of the structure of 'objects' (e.g. equivalent of fields size declarations, but better defined behaviour on fields that cross byte or word boundaries etc - see "record representation clause" in Ada, but also pay attention to representation clauses in general).

**Avoid keyword duplication**

I.e. Don't use the same keyword with different semantics depending on context

**Avoid too many symbols & their reuse**

Don't just use &, *, && all over the place 'because that's how C++ does it'. 

Also, as mentioned elsewhere, fix the '=' / '==' nonsense by 1) defining a new assignment operator, 2) use '=' to mean 'is equal to' and 3) don't allow assignment in conditions

Make short-circuiting conditions explicit (see Ada's 'and then' and 'or else' conditions) 

**Make Bool a first-class type**

Do not allow implicit 0 -> false, non-zero -> true conversion. 

There are many more, but.. 

Ultimately, every single point I've mentioned here has been about making life easier for the programmer, and making it harder for them to screw up with common mistakes. If compilers developed 40 years ago are capable of all this, modern compilers and languages should be providing core features to achieve this; it shouldn't be the responsibility of a 'template library' to do that.

**Comment by jonmeow** [comment]
We don't want our forums to be used for language debates. You can praise features of Ada without insulting C++, and when you use aggressive words, that doesn't belong in our forums. For reference, please see the code of conduct. https://github.com/carbon-language/carbon-lang/blob/trunk/CODE_OF_CONDUCT.md

Our team is aware of Ada. I think that most of your comments reflect areas that we haven't fully designed out, but I'll give a few quick comments.

Choice types are likely going to be provided instead of enum types, and they'll have different flexibilities. We're thinking through array and slice design to make them easy to use and efficient. We're trying to be conservative on casts to prevent errors, but I don't know which way int->bool casts are heading off the top of my head.

However, bear in mind we're trying to convert C++ use-cases. A lot of C++ developers are very comfortable with =/==, and there's a lot of good reasons to support assignment in conditionals in order to limit variable scopes. The nuances of syntax and error detection may be what matters more than Ada's specific choices in these areas.

**Reply by jmccabe** [reply]
Jon

Can you please specify where I've used "aggressive words"? Having re-read my comment, unless you've edited it, there are no words in there that I consider aggressive, so your help would be appreciated. 

I'd be interested in hearing more about what you mean by "choice types". It's worth remembering that, even in Ada, enumerated types are a fundamental type and are there because there is a genuine need to be able to specify and use that kind of thing. Your brief description of "choice types" makes me think you're going to end up with something that's more flexible, extendable etc etc, but more complicated (i.e. hampers the development of tools that help prove your implementation is safe) and less performant.

I am pleased your team are aware of Ada but, if it hasn't already been done, please take the time to read through and understand its history, its requirements, its design and its development before continuing to plough on and reinvent C/C++/Java/C# etc with slightly different syntax. Enumerated types are a prime example of where it's easy to see how you may be able to improve things but, in doing so, you could quite easily compromise the ease of use and performance of your language. 

FWIW, at this point I'm probably going to step back as, TBH, this whole project just appears to be Google trying to detract from Rust (and, at this point, we really need less language fragmentation, not more) but, before that, I just need to reiterate something I've said on another discussion. I've been developing software professionally since 1987. In that time, I've seen languages come and go, but I've also seen that, for those languages that have managed to keep going, the time from inception to becoming fully accepted and 'mainstream' is in excess of 10 years. From my point of view, that means that a 'new' language should be aiming to make it easy for _future_ developers to write safe, robust, understandable and performant code so justifying design choices on what _current_ developers are comfortable with, even when there are numerous, known, good reasons to avoid the same mistakes, is complete anathema to me.

Thank you for your time.

**Reply by PeterNjeim** [reply]
@jonmeow Falsely accusing someone of being "aggressive" is against the CoC and should result in a punishment. Don't water down serious topics.

By falsely accusing @jmccabe of being "aggressive", you violated the following:

- be welcoming,
- be friendly and patient,
- be considerate,
- be kind,
- be careful in the words that we choose,

The CoC recommends in this instance: _Everyone is encouraged to speak up in response to any behavior that they deem inappropriate, threatening, offensive, or harmful. If you believe someone is violating the code of conduct, please report it to the [conduct team](https://github.com/carbon-language/carbon-lang/blob/trunk/CODE_OF_CONDUCT.md#conduct-team)._

I have reported this incident to the Conduct team.

**Comment by L4stR1t3s** [comment]
A core feature of C/C++ is the ability to override memory allocation. A lot of the things you suggest make that impossible or extremely complicated. Why are all primitive data type sizes multiples of 1? Because memory allocation then requires a single unsigned integer parameter. You can't go and change this without creating a fundamentally very different language that can no longer be used for the purpose it was originally intended for and is still used today. If Carbon does not allow overriding memory allocation, it's basically dead on arrival as a C++ replacement.

STL containers have specific use cases, std::array is not a let down, it does exactly what it is meant to do. std::array is designed be usable at compile time, where a lot more limitations apply than at run-time. If it did what you want, then it'd be like other STL containers that already exist, and no STL container would exist for the specific purpose std::array exists. That would be a let down.

STL containers offer a wide range of generic use cases, for example, you can use std::map as a non-zero based indexed container. The first and last element are the lowest and highest index. Sure, it doesn't constrain what elements you can put in it for example, but that is not exactly a generic use case anymore then. Write your own implementation, or wrapper around an existing container where appropriate/needed.

**Reply by jmccabe** [reply]
Your first point is either a misunderstanding of the point, or just plain wrong. Refer to the description of Storage Pools here - https://www.adaic.org/resources/add_content/standards/05aarm/html/AA-13-11.html - which aren't dissimilar to C++'s support for overriding new and delete (assuming that's what you're referring to). In addition, take a look through [this paper](https://www.adaic.org/resources/add_content/docs/craft/html/ch05.htm) to understand the justification behind allowing user-defined types whose ranges aren't restricted to those of byte-multiple values. Of course these types have underlying types that DO conform to the machine architecture, but the developer doesn't need to know, nor care about that and:

1. Support from the compiler for these types needs to be developed _once_
2. Developers do not need to craft their own code for (unsigned) integers to wrap; the compiler handles that under the hood which tends to help avoid the potential for errors

Your 2nd point; what I said is my opinion. The C++ team could have been far more ambitious with std::array, but they restricted its scope to something that's only a little more useful than a normal C++ (C) array. Your opinion, as stated here, clearly differs.

Your 3rd point highlights the point I'm trying to make; "write your own", when you can have a fundamental container type defined by the language that does exactly what you want, is available to everyone using the language, which can be proven to be effective, efficient and safe, is just like returning to the dark ages.

In addition:

1.  I presume you do realise that your suggestion of using a map as a non-zero-based indexed container adds a performance overhead that, in many systems, is unacceptable?
3. Where do you get the idea that constraining elements, and the number of elements, is not a generic use case any more? It's certainly a use-case in C++ (which I'm hearing a lot of in the discussions for Carbon), and the introduction/implementation of std::array supports that view
4. If you look at the documentation on [Ada Containers](Ada Containers) you can see that it is perfectly valid and acceptable to have generic container classes, while also supporting arrays as a first-class type (which helps with performance)

Essentially your 3rd point, in particular, makes it the developer's responsibility (may of them) to implement what have been fundamental, efficient structures for over 50 years, potentially leading to error-prone, and bloated code. In what way is that progress?

**Reply by L4stR1t3s** [reply]
"arent' dissimilar" is a good indication that you know that ADA allocator pools are not the same.

Have you written compile-time evaluated code? I don't think so. So, because you don't write compile-time evaluated code, no one should? Seriously, if std::array did more than it does, it couldn't be used for that SPECIFIC PURPOSE IT WAS DESIGNED FOR. There is std::list, std::vector, ... use that for what you want.

std::map overhead that is unacceptable? Stop making assumptions. Do you even know how it's implemented? Yea, you can make an implementation that will be faster, but you can do that for every specific usage of a generic container. That's my entire point, write your own when needed/required.

std::array is constrained because it has to be to be used at compile time. It'd be nice to have dynamic containers at compile time, maybe some day we will, but right now we don't. When that's possible, std::array will no longer be constrained. So no, std::array is not an indication there is a need for constraints, std::array is an indication that there are constraints at compile-time that don't exist at run-time.

A container offers a specific method of storing, accessing, changing, inserting, moving, ... data and not necessarily all of those. There are specific patterns that allow for faster access, faster inserting, ... You pick the one that fits your use case the best. How would constraining indices make a difference on how a container stores data? It doesn't. Therefor you can implement a constrained version of every single container that already exists. Just like you could make a version of each container that automatically stores 10 instances of every element you add. That doesn't make it a new container. It's an abstraction of existing containers. An implementation detail, not a generic design.

Your solution to error-prone and bloated code is moving all of it to the standard library? That's a great idea! /s

**Reply by jmccabe** [reply]
I'm going to stop there; your ranting and shouting adds nothing to the discussion.

**Comment by cjdb** [comment]
I have locked this conversation, as it uses both aggressive language, and has started a language debate, neither of which we wish to cultivate in this community. For reference, please see the [code of conduct](https://github.com/carbon-language/carbon-lang/blob/trunk/CODE_OF_CONDUCT.md).

Examples of aggressive language:

> I find it shocking that the C++ community seems to be so happy to get half-baked new features added to the language rather than sorting out the fundamental inadequacies of the core language.

We don't allow jokes or generic complaints about programming languages. Those are off topic and make our community less friendly.

There are obviously changes that we'd like to make to C++, but calling C++ "half-baked" is demeaning to those who have invested years of work into that project.

> TBH, this whole project just appears to be Google trying to detract from Rust

This is disrespectful to everyone working on the Carbon Language project. While we are open to criticism of the design of the language, it isn't okay to question people's motivations. Both the Carbon Language [README](https://github.com/carbon-language/carbon-lang#why-build-carbon) and [FAQ](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/faq.md#if-you-can-use-rust-ignore-carbon) state that if an existing programming language meets your needs, then you should ignore Carbon and use that instead (with the latter explicitly addressing Rust).
